### ISSUE TITLE
encoding/json: cannot call UnmarshalJSON on value receiver through interface

### ISSUE THREAD

**Issue by dsnet** [golang/go#27722]
[Noticed this inconsistency when poking around the code](https://play.golang.org/p/Yw_2jS_Kr6r):
```go
type CustomUnmarshaler struct{ p *bytes.Buffer }

func (cu CustomUnmarshaler) UnmarshalJSON(b []byte) error {
	cu.p.Write(b)
	return nil
}
func (cu CustomUnmarshaler) String() string {
	return fmt.Sprintf("custom %v", cu.p.String())
}

func main() {
	s := &struct {
		F1 CustomUnmarshaler
		F2 interface{}
		F3 interface{}
	}{
		CustomUnmarshaler{new(bytes.Buffer)},
		CustomUnmarshaler{new(bytes.Buffer)},
		&CustomUnmarshaler{new(bytes.Buffer)},
	}
	json.Unmarshal([]byte(`{"F1": "F1", "F2": "F2", "F3": "F3"}`), s)
	fmt.Println(s.F1)
	fmt.Println(s.F2)
	fmt.Println(s.F3)
}
```

This current prints:
```
custom "F1"
F2
custom "F3"
```

I expect it to print:
```
custom "F1"
custom "F2"
custom "F3"
```

In this case, `UnmarshalJSON` is a method on the value receiver. This is a rare situation for methods that mutate the receiver, but is valid. I expect the method to still be called. That is, the `json` package should not assume that the receiver must be a pointer.

This is a very obscure use-case and I personally don't have a horse in the game to address this.

**Comment by mvdan** [comment]
In the pointer case, is pre-existing data kept? If I understand addressability correctly, in the non-pointer interface field we cannot modify the existing data in-place.

Perhaps we could simply create a new value via reflection, but if the pointer field doesn't get zeroed before decoding, that would be inconsistent.

**Comment by dsnet** [comment]
> In the pointer case, is pre-existing data kept?

Yes. The behavior of `json.Unmarshal` is to merge into pre-allocated data structures. It is a slight inconsistency that this does not happen here.

**Comment by bcmills** [comment]
CC @rsc @bradfitz

**Comment by rsnancollas** [comment]
It looks like a solution was proposed, but never got merged in. I ran into this bug last night and it would be great to fix it.

**Comment by dsnet** [comment]
The v1 behavior is more or less frozen for compatibility reasons. With the release of Go 1.27, json/v2 now consistently always call methods on pointer receivers even if it only has a value receiver on hand.
